miasm/expression/simplifications_common.py

```diff
@@ -1753,19 +1753,39 @@ def simp_compose_and_mask(_, expr):
     if not arg2.is_int():
         return expr
     int2 = int(arg2)
-    if (int2 + 1) & int2 != 0:
-        return expr
-    mask_size = int2.bit_length() + 7 // 8
+    mask_size = (int2.bit_length() + 7) // 8 * 8
```

Comment:
I don't understand why is there a +7 here
Can you

Reply:
the mask size is rounded up to the next multiple of 8. i think this is not actually needed, you could just use `int2.bit_length()` instead, but this may result in more complex expressions because of this line in the loop:

```
            if arg.is_int() and offset < mask_size < offset+arg.size:
                arg = ExprSlice(arg, 0, mask_size-offset)
```
this might be the result without rounding up:
```
{X 0 8, Y 8 16} & 0x7FFF -> {x 0 8, Y & 0x7F 8 15, 0 15 16}
```
with rounding up:
```
{X 0 8, Y 8 16} & 0x7FFF -> {X 0 8, Y 8 16} & 0x7FFF
```

Reply:
Ok. The trick is that in Miasm compose can have, and *have* components of differents sizes. Not only 8 bit arguments.
So maybe we will have a problem here. I think we can distinguish 2 big cases.
- The first one, in which this simplification will be very interesting is the case where we make elements "disappear". (the example you gave)
- The second one, in which we will propagate the mask *into* the compose, and this propagation will only create multiple new mask for each (or most) sub arguments. In this case, maybe we don't want to propagate the mask

So maybe we can use this metric to do or not, this simplification
So the code may be:
- do the simplfication as you propose
- once it's done, see if the number of ExprCompose arguments is reduced (compared to the original one).
- if yes, return this expression
- if no, return the original expression
This way, the simplification should not generate a "more complex" expression (in term of number of operations)

I am aware of the fact that it may do some work for nothing, but hey, the 'simplification' problem is hard :smile: 

What do you think of this proposition?

Reply:
i think this is a good idea, but it would be better to check if the resulting `ExprCompose` is simpler than the old one. i.e. if parts of the expression were replaced by constants.